tests/test_search.py

```diff
@@ -304,3 +304,23 @@ def test_parallel(app):
     app.build()
     index = load_searchindex(app.outdir / 'searchindex.js')
     assert index['docnames'] == ['index', 'nosearch', 'tocitem']
+
+
+@pytest.mark.sphinx(testroot='search')
+def test_search_index_is_deterministic(app):
+    app.builder.build_all()
+    index = load_searchindex(app.outdir / 'searchindex.js')
+
+    def check_if_dicts_are_sorted(item):
+        if isinstance(item, dict):
+            assert list(item.keys()) == sorted(item.keys())
+            for child in item.values():
+                check_if_dicts_are_sorted(child)
+        # Lists in the search index cannot be sorted: for some lists, their
+        # position inside the list is referenced elsewhere in the index, so if
+        # we were to sort lists, the search index would break.
```

Comment:
I don't see how it's possible to sort all lists without changing the format of the search index itself, which feels outside the scope of this PR. Those orderings are deterministic, just not sortable.

There are two cases today where items are not sorted inside of lists:

* In the `.titles` and `.filenames` lists, each item corresponds to the item at the same position in the (sorted) `.docnames` list. Sorting `.titles` and `.filenames` would break that relationship.

* In the `.alltitles.*` lists, the lists are actually an `(int, str)` tuple, which is represented as a list in JSON. Sorting them wouldn't make much sense, and won't work anyway due to the different types.

Still, I updated the test to check that everything but those two exceptions are sorted.

Reply:
> Those orderings are deterministic

As long as that's true, this seems reasonable to me.  For a moment I was worried that if `.titles` / `.filenames` were _not_ deterministic, then we'd be solving 95% of cases while leaving a more difficult-to-resolve case deeper in the datastructure.

Could I check why/how confident we are that those fields are deterministically-generated, to get a sense for how robust or fragile that is?

Reply:
> Those orderings are deterministic

Ahh, this wasn't clear in the comment -- please would you be able to update it just to clarify?

A

Reply:
> Ahh, this wasn't clear in the comment -- please would you be able to update it just to clarify?

Updated the comments to clarify why they are deterministic!

> Could I check why/how confident we are that those fields are deterministically-generated, to get a sense for how robust or fragile that is?

The code seems pretty robust:

https://github.com/sphinx-doc/sphinx/blob/fcc38997f1d9b728bb4ffc64fc362c7763a4ee25/sphinx/search/__init__.py#L385-L386

Also, there is another test that verifies the search index content is the expected one, which (implicitly) tests the relationship between docnames and titles/filenames.

https://github.com/sphinx-doc/sphinx/blob/fcc38997f1d9b728bb4ffc64fc362c7763a4ee25/tests/test_search.py#L179-L197